File: rust/solutions/src/year2023/day1.rs
use aoc_derive::{aoc, aoc_input};

#[aoc_input(day1)]
fn parse(input: &str) -> Vec<String> {
    input.lines().map(|s| s.trim().to_string()).collect()
}

#[aoc(day1, part1)]
fn part1(input: &[String]) -> u32 {
    input.into_iter().flat_map(|s| extract_numbers(s)).sum()
}

#[aoc(day1, part2)]
fn part2(input: &[String]) -> u32 {
    input
        .into_iter()
        .map(|s| convert_words_to_numbers(s))
        .flat_map(|s| extract_numbers(&s))
        .sum()
}

fn extract_numbers(s: &str) -> Result<u32, String> {
    let first_digit = s.chars().filter(|c| c.is_numeric()).next();
    let last_digit = s.chars().filter(|c| c.is_numeric()).last();

    let num = match (first_digit, last_digit) {
        (Some(n1), Some(n2)) => format!("{}{}", n1, n2),
        (_, _) => return Err("couldn't find two numbers in input string".to_string()),
    };

    num.parse()
        .map_err(|_| "unable to parse number from input string".to_string())
}

fn convert_words_to_numbers(s: &str) -> String {
    s.replace("one", "on1e")
        .replace("two", "tw2o")
        .replace("three", "thre3e")
        .replace("four", "fou4r")
        .replace("five", "fiv5e")
        .replace("six", "si6x")
        .replace("seven", "seve7n")
        .replace("eight", "eigh8t")
        .replace("nine", "nin9e")
}

#[cfg(test)]
mod tests {
    use super::*;

    const INPUT_1: &str = "1abc2
    pqr3stu8vwx
    a1b2c3d4e5f
    treb7uchet";

    const INPUT_2: &str = "two1nine
    eightwothree
    abcone2threexyz
    xtwone3four
    4nineeightseven2
    zoneight234
    7pqrstsixteen";

    #[test]
    fn test_part1() {
        let entries = parse(INPUT_1);
        assert_eq!(part1(&entries), 142);
    }

    #[test]
    fn test_part2() {
        let entries = parse(INPUT_2);
        assert_eq!(part2(&entries), 281);
    }
}

File: zig/src/y2023/day1.zig
const std = @import("std");

const numberConversion = std.StaticStringMap(u32).initComptime(.{
    .{ "0", 0 },
    .{ "1", 1 },
    .{ "2", 2 },
    .{ "3", 3 },
    .{ "4", 4 },
    .{ "5", 5 },
    .{ "6", 6 },
    .{ "7", 7 },
    .{ "8", 8 },
    .{ "9", 9 },
    .{ "zero", 0 },
    .{ "one", 1 },
    .{ "two", 2 },
    .{ "three", 3 },
    .{ "four", 4 },
    .{ "five", 5 },
    .{ "six", 6 },
    .{ "seven", 7 },
    .{ "eight", 8 },
    .{ "nine", 9 },
});

pub fn run(input: []const u8) !void {
    const answer_part1: u64 = try part1(input);
    const answer_part2: u64 = try part2(input);

    std.log.info("Day 1", .{});
    std.log.info("  Part 1: {}", .{answer_part1});
    std.log.info("  Part 2: {}\n", .{answer_part2});
}

fn part1(input: []const u8) !u64 {
    var lines = std.mem.tokenizeScalar(u8, input, '\n');
    var sum: u64 = 0;

    while (lines.next()) |line| {
        var i: usize = 0;
        var first_digit: u32 = undefined;
        while (i < line.len) : (i += 1) {
            if (std.ascii.isDigit(line[i])) {
                first_digit = numberConversion.get(line[i .. i + 1]).?;
                break;
            }
        }

        var j: usize = line.len - 1;
        var second_digit: u32 = undefined;
        while (j >= 0) : (j -= 1) {
            if (std.ascii.isDigit(line[j])) {
                second_digit = numberConversion.get(line[j .. j + 1]).?;
                break;
            }
        }

        sum += (first_digit * 10) + second_digit;
    }

    return sum;
}

fn part2(input: []const u8) !u64 {
    var lines = std.mem.tokenizeScalar(u8, input, '\n');
    var sum: u64 = 0;

    while (lines.next()) |line| {
        var i: usize = 0;
        var first_digit: u32 = undefined;
        search_loop: while (i < line.len) : (i += 1) {
            for (numberConversion.keys()) |key| {
                if (std.mem.startsWith(u8, line[i..line.len], key)) {
                    first_digit = numberConversion.get(key).?;
                    break :search_loop;
                }
            }
        }

        var j: usize = line.len;
        var second_digit: u32 = undefined;
        search_loop: while (j > 0) : (j -= 1) {
            for (numberConversion.keys()) |key| {
                if (std.mem.endsWith(u8, line[0..j], key)) {
                    second_digit = numberConversion.get(key).?;
                    break :search_loop;
                }
            }
        }

        sum += (first_digit * 10) + second_digit;
    }

    return sum;
}
